libs/ast_generic/AST_generic.ml

```diff
@@ -2405,3 +2395,48 @@ let is_metavar_name s = Common.( =~ ) s "^\\(\\$[A-Z_][A-Z_0-9]*\\)$"
 (* coupling: Metavariable.is_metavar_ellipsis *)
 let is_metavar_ellipsis s =
   Common.( =~ ) s "^\\(\\$\\.\\.\\.[A-Z_][A-Z_0-9]*\\)$"
+
+(*****************************************************************************)
+(* Custom visitors *)
+(*****************************************************************************)
+
+(* Most clients should use this instead of the default `iter`. In many cases,
+ * it's not desirable to recurse into id_info since it contains resolved names
+ * and svalues which often contain nodes that are already present elsewhere in
+ * the AST. This matches the default behavior of the old mk_visitor. *)
+class virtual ['self] iter_no_id_info =
+  object (_self : 'self)
+    inherit ['self] iter
+    method! visit_id_info _env _info = ()
+  end
+
+(* This is based on the old behavior of Map_AST.mk_visitor. If you are adding a
+ * new callsite, you should consider whether the nonstandard behavior here makes
+ * sense for your use case. Think about using the autogenerated AST_generic.map
+ * directly. *)
+class virtual ['self] map_custom =
+  object (self : 'self)
+    inherit [_] map
+
+    method! visit_tok _env v =
+      (* Make a copy, since Tok.t is mutable *)
+      let v = { v with Tok.token = v.token } in
+      v
+
+    method! visit_expr env x =
+      let ekind = self#visit_expr_kind env x.e in
+      (* TODO? reuse the e_id or create a new one? *)
+      e ekind
```

Comment:
This calls `e`, so I had to put it below the definition. I moved `iter_no_id_info` too, for consistency.